File: scripts/hello-morph/docker-buildkit-example/start_snapshot.py
# snapshot_2lj87jj7

from dotenv import load_dotenv
from morphcloud.api import MorphCloudClient

load_dotenv()


client = MorphCloudClient()

snapshot_id = "snapshot_ojixukjb"
# snapshot_id = "snapshot_2lj87jj7"

instance = client.instances.start(snapshot_id=snapshot_id)
print(f"Created instance: {instance.id}")

instance.expose_http_service("openvscode", 2376)
instance.expose_http_service("worker", 3002)
instance.expose_http_service("management", 3003)

for service in instance.networking.http_services:
    if service.name == "openvscode":
        print(f"- OpenVSCode: {service.url}/?folder=/root/workspace")
        continue
    print(f"- {service.name}: {service.url}")

File: scripts/hello-morph/docker-buildkit-example/start_snapshot.ts
import type {
  ServerToWorkerEvents,
  WorkerToServerEvents,
} from "@coderouter/shared";
import { MorphCloudClient } from "morphcloud";
import { io, Socket } from "socket.io-client";

const client = new MorphCloudClient();
const instance = await client.instances.start({
  snapshotId: "snapshot_gn1wmycs",
});

console.log(`Created instance: ${instance.id}`);

const exposedServices = instance.networking.httpServices;
const vscodeService = exposedServices.find((service) => service.port === 2376);
const workerService = exposedServices.find((service) => service.port === 3002);
if (!vscodeService || !workerService) {
  throw new Error("VSCode or worker service not found");
}

console.log(`VSCode: ${vscodeService.url}/?folder=/root/workspace`);

// connect to the worker with socketio
const clientSocket = io(workerService.url, {
  timeout: 10000,
  reconnectionAttempts: 3,
}) as Socket<WorkerToServerEvents, ServerToWorkerEvents>;

clientSocket.on("connect", () => {
  console.log("Connected to worker");
  // clientSocket.emit("get-active-terminals");
  // dispatch a tack
  clientSocket.emit("worker:create-terminal", {
    terminalId: crypto.randomUUID(),
    cols: 80,
    rows: 24,
    cwd: "/root/workspace",
    command: "bun x opencode-ai 'whats the time'",
  });
});

clientSocket.on("disconnect", () => {
  console.log("Disconnected from worker");
});

process.on("SIGINT", async () => {
  console.log("Stopping instance");
  await instance.stop();
  process.exit(0);
});

console.log("Press Ctrl+C to stop instance");
await new Promise(() => void {});
